### PR TITLE
WebGPU: remove rtt encoder + misc changes

### PR DESCRIPTION
**Pull request by Popov72**
The main purpose of the PR is to remove the renderTargetEncoder and keep only two encoders, "copy" and "render".

It simplifies the code and fixes some problems (see the two visualization tests I added - they don't work as expected without the PR).

I took this opportunity to rework some parts of the code to simplify it further.

Also:
* I had to disable the "Instances + GBR + motion blur" visualization test because Chrome now reports an error for it ("Total color attachment bytes per sample (48) exceeds maximum (32)"). We will have to wait for an extension that allows to raise this  limit before re-enabling it.
* With the latest version of Chrome (119), I also had to handle depth textures in much the same way as float textures when linear filtering is not supported for them: the sampler type must be forced to "unfilterable float" for depth textures in the bind group layout.
* The "float32-filtering" extension is now supported by Chrome, which revealed some bugs I had to fix in the code (I had to create a new "supportFloatTexturesResolve" caps because WebGPU can currently render to a R32float but not resolve it).
* I fixed a bug with greased lines, which didn't work in WebGPU.
* I've added the visualization tests I used specifically for WebGPU to the regular list (in `config.json`), as it's no problem to have them for WebGL too.

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14506/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14506/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14506/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14506/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14506/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14506/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14506/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14506/merge#BCU1XR#0

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14506/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for webgl2 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14506/merge/testResults/webgl2/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by Popov72** [comment]
@RaananW All visu tests that fail in WebGL work locally. It seems it's a problem of timing, they all have a renderCount > 0 to take a meaningful frame...

**Comment by bjsplat** [comment]
Visualization tests for webgl2 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14506/merge/testResults/webgl2/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14506/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by RaananW** [comment]
I'll check that tomorrow

File: packages/dev/core/src/Engines/webgpuEngine.ts [hunk 1]
```diff
@@ -85,6 +81,16 @@ const disableUniformityAnalysisMarker = "/* disable_uniformity_analysis */";
 
 const tempColor4 = new Color4();
 
+/** @internal */
+class WebGPURenderPassWrapper {
```
Comment (sebavan) [inline packages/dev/core/src/Engines/webgpuEngine.ts:85]: This should be an interface or type only I do not thing we need a class here ?

File: packages/dev/core/src/Rendering/prePassRenderer.ts [hunk 2]
```diff
@@ -284,10 +284,16 @@ export class PrePassRenderer {
             type = Constants.TEXTURETYPE_HALF_FLOAT;
         }
 
-        if (type !== Constants.TEXTURETYPE_FLOAT) {
-            for (let i = 0; i < PrePassRenderer.TextureFormats.length; ++i) {
-                if (PrePassRenderer.TextureFormats[i].type === Constants.TEXTURETYPE_FLOAT) {
-                    PrePassRenderer.TextureFormats[Constants.PREPASS_DEPTH_TEXTURE_TYPE].type = type;
+        for (let i = 0; i < PrePassRenderer.TextureFormats.length; ++i) {
+            const format = PrePassRenderer.TextureFormats[i].format;
+            if (PrePassRenderer.TextureFormats[i].type === Constants.TEXTURETYPE_FLOAT) {
```
Comment (sebavan) [inline packages/dev/core/src/Rendering/prePassRenderer.ts:289]: Isnt all of this only changing on TextureFormats changes ? (wonder if we could cache all the conditions)

**Review by sebavan — APPROVED** [review]
All good but a couple neats.

How is the perf impact on the shader compilation (findbackward function) ?

File: packages/dev/core/src/Rendering/prePassRenderer.ts (hunk 2 above)
Comment (Popov72) [reply packages/dev/core/src/Rendering/prePassRenderer.ts:289]: This code is done only once, in the constructor, to account for the capability of the current hardware. I'm not sure what you mean by caching the conditions?

**Comment by Popov72** [comment]
> How is the perf impact on the shader compilation (findbackward function) ?

I don't think it will make a difference because:
* it is called only for functions that must be inlined (there are only a few in all our shader code)
* it is called only for inlined functions that don't return void
* the `do...while` loop which is looking for a "\n" or a "{" will typically run only a few iterations to find one of these characters

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14506/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for webgl2 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14506/merge/testResults/webgl2/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by Popov72** [comment]
@RaananW Isn't the flag to remove a test from the CI named `excludeFromAutomaticTesting`? I added it to the 5 tests that failed and they still run...

File: packages/tools/tests/test/visualization/config.json [hunk 3]
```diff
@@ -1778,6 +1778,198 @@
...
+        },
+        {
+            "title": "soft-transparent-shadows",
+            "playgroundId": "#G3DJGA#0",
+            "excludedEngines": ["webgl1"]
+        },
+        {
+            "title": "default-render-pipeline",
+            "playgroundId": "#GAFR36#0",
+            "renderCount": 5
```
Comment (RaananW) [inline packages/tools/tests/test/visualization/config.json:1847]:
```suggestion
            "renderCount": 5,
            "excludeFromAutomaticTesting": true
```

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14506/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

File: packages/tools/tests/test/visualization/config.json [hunk 4]
```diff
@@ -1778,6 +1778,199 @@
...
+        },
+        {
+            "title": "change-texture-of-material",
+            "renderCount": 10,
+            "playgroundId": "#7PN9PH#2"
+        },
+        {
+            "title": "screen-space-curvature",
+            "renderCount": 5,
+            "playgroundId": "#YF8D42#22"
```
Comment (RaananW) [inline packages/tools/tests/test/visualization/config.json:1961]:
```suggestion
            "playgroundId": "#YF8D42#22",
            "excludedEngines": ["webgl1"]
```

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14506/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for webgl2 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14506/merge/testResults/webgl2/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14506/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14506/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by Popov72** [comment]
I don't know what the problem is with the "change-texture-of-material" in WebGL1, as it does work locally...
It's not that important, so I disabled it.

**Comment by bjsplat** [comment]
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14506/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

**Comment by RaananW** [comment]
the test Command Order WebGPU 1 fails in webgl 1 - `uncaught exception: HDR prefiltering is not available in WebGL 1., you can use real time filtering instead.`. Maybe just filter it our of webgl1?

**Pull request merged**